config/crd/bases/metal3.io_baremetalhosts.yaml

```diff
@@ -121,6 +121,10 @@ spec:
               description:
                 description: Description is a human-entered text used to help identify the host
                 type: string
+              disableAutomatedClean:
```

Comment:
I know the design for this landed already, but are we OK with using an annotation for disabling inspection ref metal3-io/metal3-docs#155 and #787, but disabling cleaning via the BMH spec?

Reply:
@dhellmann any thoughts on that?  I'm not blocking on it, just wondering if we have any guidelines on what should be in the spec vs annotations?

In practice it seems like annotations are used to modify the baremetalhost_controller behavior, but the spec is used to influence the provisioner configuration, but that seems like an implementation detail rather than an API design pattern :)

Reply:
I'm not sure why we used an annotation for disabling inspection. Maybe because it wasn't really a thing we thought most users would (or should) do? Or because the semantics of changing an API field to enable inspection after a host was provisioned were confusing?

An API field to turn off cleaning seems reasonable, and the semantics of changing it later are clearer (at least to me). I could be convinced that the inspection and cleaning features should be treated the same way, as API fields.

I should have provided this feedback on the design review, but I think we don't want to use a boolean field. It's harder to manage changes in behavior. For example, today cleaning is all or none. In the future, it may be possible to clean some disks and leave others alone. We may want a 3rd mode to cover that case. Creating this field as an enum now would let us make that change relatively easily, without an API version change and upgrade process. So, I suggest a name like `automatedCleaningMode` with values `Full` and `None` instead of a bool here.

Reply:
> I'm not sure why we used an annotation for disabling inspection.

I did it that way to align with the re-introspection proposal that was already accepted : https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md

> Maybe because it wasn't really a thing we thought most users would (or should) do? 

This is probably also true, the main use-case we have for it atm is when booting a live-iso, which likely isn't the most common path, and we might encourage folks to use a visible API in the spec for the wrong reasons (e.g faster deployments, but the same could be said for this API I guess)

Reply:
> > I'm not sure why we used an annotation for disabling inspection.
> 
> I did it that way to align with the re-introspection proposal that was already accepted : https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md

Ah, good point. In that case the annotation is removed after the re-inspection is done, right? I suppose we could have had the operator toggle a flag that's part of the schema instead. But we probably based the re-inspection thing on the reboot API, and that was also based on an annotation.

> > Maybe because it wasn't really a thing we thought most users would (or should) do?
> 
> This is probably also true, the main use-case we have for it atm is when booting a live-iso, which likely isn't the most common path, and we might encourage folks to use a visible API in the spec for the wrong reasons (e.g faster deployments, but the same could be said for this API I guess)

Yeah, true. I have seen recommendations for using annotations for interactions between components (such as an annotation on a host that is consumed by something other than the BMO) or as early implementations of APIs that aren't ready to be locked into a CRD. That last use feels vague and fuzzy, but I think that's how I justified the reboot, re-inspection, and skip-inspection changes in my mind. Eventually, those *should* become part of the spec fields, when we figure out how to do that. This change seems to fit easily enough into the API fields that we can move right to that implementation. But I'll go with the group's consensus, so if others would be more comfortable using an annotation, that's OK with me.

Reply:
> > > I'm not sure why we used an annotation for disabling inspection.
> > 
> > 
> > I did it that way to align with the re-introspection proposal that was already accepted : https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md
> 
> Ah, good point. In that case the annotation is removed after the re-inspection is done, right? 

That's right. The implementation PR is not merged yet though https://github.com/metal3-io/baremetal-operator/pull/607

>I suppose we could have had the operator toggle a flag that's part of the schema instead. But we probably based the re-inspection thing on the reboot API, and that was also based on an annotation.
> 
That's true. Re-inspection was designed to be similar interface as reboot API.

Reply:
> I should have provided this feedback on the design review, but I think we don't want to use a boolean field. It's harder to manage changes in behavior. For example, today cleaning is all or none. In the future, it may be possible to clean some disks and leave others alone. We may want a 3rd mode to cover that case. Creating this field as an enum now would let us make that change relatively easily, without an API version change and upgrade process. So, I suggest a name like `automatedCleaningMode` with values `Full` and `None` instead of a bool here. We can update the design doc after the code is approved.

@dhellmann Thanks, I agree with your point here and can change the both code + proposal. But are we okay to have enum here considering Ironic `Node.AutomatedClean` is bool. I know `automatedCleaningMode` doesn't have to be the same type as  `Node.AutomatedClean`, but just asking if there could be any issue with that.

Reply:
> But I'll go with the group's consensus, so if others would be more comfortable using an annotation, that's OK with me.

+1 I'm fine with this going into the spec, I just wanted to ensure we're all OK with that approach :)

And I agree that in the case of the inspection API it's fine to use an annotation for consistency (it'd be confusing to have both annotations and spec fields modifying that behavior IMO), and that's definitely an API which might see further changes in future, so it makes sense to avoid committing to a spec API now.

Reply:
> @dhellmann Thanks, I agree with your point here and can change the both code + proposal. But are we okay to have enum here considering Ironic `Node.AutomatedClean` is bool. I know `automatedCleaningMode` doesn't have to be the same type as `Node.AutomatedClean`, but just asking if there could be any issue with that.

I think it's OK for the APIs to be different. We've never made the assumption that metal3 users are Ironic users, and we try to design our APIs following kubernetes standards and conventions.